prov/util/src/util_ns.c

```diff
@@ -0,0 +1,598 @@
+	memcpy(name, name_in, ns->name_len);
+
+	if (rbtFind(ns->ns_map, service)) {
+		ret = -FI_EADDRINUSE;
+		goto err3;
+	}
+
+	if (rbtInsert(ns->ns_map, service, name)) {
+		ret = -FI_ENOMEM;
+		goto err3;
+	}
+	return FI_SUCCESS;
+
+err3:
+	free(name);
+err2:
+	free(service);
+err1:
+	return ret;
+}
+ 
+static int util_ns_map_del(struct util_ns *ns, void *service_in,
+			   void *name_in)
+{
+	RbtIterator it;
+	int ret = -FI_ENOENT;
+	void *service, *name;
+
+        it = rbtFind(ns->ns_map, service_in);
+        if (it) {
+		rbtKeyValue(ns->ns_map, it, &service, &name);
+		if (memcmp(name, name_in, ns->name_len))
+			return ret;
+		free(service);
+		free(name);
+		rbtErase(ns->ns_map, it);
+		ret = FI_SUCCESS;
+	}
+
+	return ret;
+}
+
+static int util_ns_map_lookup(struct util_ns *ns, void *service_in,
+			      void **name_out)
+{
+	RbtIterator it;
+	void *key, *name;
+
+        it = rbtFind(ns->ns_map, service_in);
+	if (!it)
+		return -FI_ENOENT;
+
+	rbtKeyValue(ns->ns_map, it, &key, (void **)&name);
+	*name_out = name;
+
+	if (ns->is_service_wildcard && ns->is_service_wildcard(service_in))
+		memcpy(service_in, key, ns->service_len);
+
+	return FI_SUCCESS;
+}
+
+static void util_ns_name_server_cleanup(void *args)
+{
+	void **cleanup_args = (void **)args;
+	ofi_close_socket((uintptr_t)cleanup_args[0]);
+	util_ns_map_fini((struct util_ns *)cleanup_args[1]);
+}
+
+static int util_ns_op_dispatcher(struct util_ns *ns,
+				 struct util_ns_cmd *cmd,
+				 SOCKET sock)
+{
+	int ret = FI_SUCCESS;
+	size_t io_len = 0;
+	void *io_buf = NULL, *service, *name;
+
+	switch (cmd->op) {
+	case OFI_UTIL_NS_ADD:
+	case OFI_UTIL_NS_DEL:
+		io_len = ns->name_len + ns->service_len;
+		io_buf = calloc(io_len, 1);
+		if (!io_buf) {
+			ret = -FI_ENOMEM;
+			goto fn1;
+		}
+
+		ret = util_ns_read_socket_op(sock, io_buf, io_len);
+		if (ret == io_len) {
+			service = io_buf;
+			name = (void *)((char *)io_buf + ns->service_len);
+			ret = (cmd->op == OFI_UTIL_NS_ADD) ?
+			      util_ns_map_add(ns, service, name) :
+			      util_ns_map_del(ns, service, name);
+		} else {
+			ret = -FI_ENODATA;
+		}
+		goto fn2;
+
+	case OFI_UTIL_NS_QUERY:
+		io_len = ns->service_len;
+		/* allocate io_buf to be large enough */
+		service = io_buf = calloc(
+			cmd_len + ns->service_len + ns->name_len, 1
+		);
+		if (!io_buf) {
+			ret = -FI_ENOMEM;
+			goto fn1;
+		}
+
+		ret = util_ns_read_socket_op(sock, io_buf, io_len);
+		if (ret == io_len) {
+			cmd->op = OFI_UTIL_NS_ACK;
+			cmd->status = util_ns_map_lookup(
+				ns, service, &name
+			);
+		} else {
+			ret = -FI_ENODATA;
+			goto fn2;
+		}
+
+		if (!cmd->status) {
+			io_len = 0;
+			memcpy(io_buf, cmd, cmd_len);
+			io_len += cmd_len;
+			memcpy((void *)((char *)io_buf + io_len), service,
+			       ns->service_len);
+			io_len += ns->service_len;
+			memcpy((void *)((char *)io_buf + io_len), name,
+			       ns->name_len);
+			io_len += ns->name_len;
+		} else {
+			memcpy(io_buf, cmd, cmd_len);
+			io_len = cmd_len;
+		}
```

Comment:
The "if ... else ..." section can now be simplified:
if (!cmd->status)
        io_len = cmd_len + ns->service_len + ns->name_len;
else
        io_len = cmd_len;